File: Java/libraries/recognizers-text-choice/src/main/java/com/microsoft/recognizers/text/choice/extractors/ChoiceExtractor.java
package com.microsoft.recognizers.text.choice.extractors;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.IExtractor;
import com.microsoft.recognizers.text.choice.utilities.UnicodeUtils;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import com.microsoft.recognizers.text.utilities.StringUtility;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;

public class ChoiceExtractor implements IExtractor {

    private IChoiceExtractorConfiguration config;

    public ChoiceExtractor(IChoiceExtractorConfiguration config) {
        this.config = config;
    }

    @Override
    public List<ExtractResult> extract(String text) {

        List<ExtractResult> results = new ArrayList<>();
        String trimmedText = text.toLowerCase();
        List<ExtractResult> partialResults = new ArrayList<>();
        List<String> sourceTokens = tokenize(trimmedText);

        if (text.isEmpty()) {
            return results;
        }

        for (Map.Entry<Pattern, String> entry : this.config.getMapRegexes().entrySet()) {

            Pattern regexKey = entry.getKey();
            String constantValue = entry.getValue();
            Match[] matches = RegExpUtility.getMatches(regexKey, trimmedText);
            double topScore = 0;

            for (Match match : matches) {

                List<String> matchToken = tokenize(match.value);
                for (int i = 0; i < sourceTokens.size(); i++) {
                    double score = matchValue(sourceTokens, matchToken, i);
                    topScore = Math.max(topScore, score);
                }

                if (topScore > 0.0) {
                    int start = match.index;
                    int length = match.length;
                    partialResults.add(
                        new ExtractResult(
                            start,
                            length,
                            text.substring(start, length + start),
                            constantValue,
                            new ChoiceExtractDataResult(text, topScore, new ArrayList<>())
                        )
                    );
                }

            }
        }

        if (partialResults.size() == 0) {
            return results;
        }

        partialResults.sort(Comparator.comparingInt(er -> er.getStart()));

        if (this.config.getOnlyTopMatch()) {

            double topScore = 0;
            int topResultIndex = 0;

            for (int i = 0; i < partialResults.size(); i++) {

                ChoiceExtractDataResult data = (ChoiceExtractDataResult)partialResults.get(i).getData();
                if (data.score > topScore) {
                    topScore = data.score;
                    topResultIndex = i;
                }

            }
            results.add(partialResults.get(topResultIndex));
            partialResults.remove(topResultIndex);
        } else {
            results = partialResults;
        }

        return results;
    }

    private final double matchValue(List<String> source, List<String> match, int startPosition) {

        double matched = 0;
        double totalDeviation = 0;
        double score = 0;

        for (String token : match) {
            int pos = indexOfToken(source, token, startPosition);
            if (pos >= 0) {
                int distance = matched > 0 ? pos - startPosition : 0;
                if (distance <= config.getMaxDistance()) {
                    matched++;
                    totalDeviation += distance;
                    startPosition = pos + 1;
                }
            }
        }

        if (matched > 0 && (matched == match.size() || config.getAllowPartialMatch())) {
            double completeness = matched / match.size();
            double accuracy = completeness * (matched / (matched + totalDeviation));
            double initialScore = accuracy * (matched / source.size());
            score = 0.4 + (0.6 * initialScore);
        }

        return score;
    }

    private static int indexOfToken(List<String> tokens, String token, int startPos) {

        if (tokens.size() <= startPos) {
            return -1;
        }

        return tokens.indexOf(token);
    }

    private final List<String> tokenize(String text) {

        List<String> tokens = new ArrayList<>();
        List<String> letters = UnicodeUtils.letters(text);
        String token = "";

        for (String letter : letters) {

            Optional<Match> isMatch = Arrays.stream(RegExpUtility.getMatches(this.config.getTokenRegex(), letter)).findFirst();
            if (UnicodeUtils.isEmoji(letter)) {

                // Character is in a Supplementary Unicode Plane. This is where emoji live so
                // we're going to just break each character in this range out as its own token.
                tokens.add(letter);
                if (!StringUtility.isNullOrWhiteSpace(token)) {
                    tokens.add(token);
                    token = "";
                }

            } else if (!(isMatch.isPresent() || StringUtility.isNullOrWhiteSpace(letter))) {
                token = token + letter;
            } else if (!StringUtility.isNullOrWhiteSpace(token)) {
                tokens.add(token);
                token = "";
            }
        }

        if (!StringUtility.isNullOrWhiteSpace(token)) {
            tokens.add(token);
            token = "";
        }

        return tokens;
    }
}

File: .NET/Microsoft.Recognizers.Text.Choice/Extractors/ChoiceExtractor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Text.Choice.Utilities;

namespace Microsoft.Recognizers.Text.Choice
{
    public class ChoiceExtractor : IExtractor
    {
        private readonly IChoiceExtractorConfiguration config;

        public ChoiceExtractor(IChoiceExtractorConfiguration config)
        {
            this.config = config;
        }

        public List<ExtractResult> Extract(string text)
        {
            var results = new List<ExtractResult>();
            var trimmedText = text.ToLower();

            if (string.IsNullOrEmpty(text))
            {
                return results;
            }

            var partialResults = new List<ExtractResult>();
            var sourceTokens = Tokenize(trimmedText);

            foreach (var item in config.MapRegexes)
            {
                var matches = item.Key.Matches(trimmedText);
                var topScore = 0.0;
                foreach (Match match in matches)
                {
                    var matchToken = Tokenize(match.Value);
                    for (int i = 0; i < sourceTokens.Count; i++)
                    {
                        var score = MatchValue(sourceTokens, matchToken, i);
                        topScore = Math.Max(topScore, score);
                    }

                    if (topScore > 0.0)
                    {
                        var start = match.Index;
                        var length = match.Length;
                        partialResults.Add(new ExtractResult()
                        {
                            Start = start,
                            Length = length,
                            Text = text.Substring(start, length).Trim(),
                            Type = item.Value,
                            Data = new ChoiceExtractDataResult()
                            {
                                Source = text,
                                Score = topScore,
                                OtherMatches = new List<ExtractResult>(),
                            },
                        });
                    }
                }
            }

            if (partialResults.Count == 0)
            {
                return results;
            }

            partialResults = partialResults.OrderBy(l1 => l1.Start).ToList();

            if (config.OnlyTopMatch)
            {
                var topScore = 0.0;
                var topResultIndex = 0;
                for (int i = 0; i < partialResults.Count; i++)
                {
                    var data = partialResults[i].Data as ChoiceExtractDataResult;
                    if (data.Score > topScore)
                    {
                        topScore = data.Score;
                        topResultIndex = i;
                    }
                }

                var topResultData = partialResults[topResultIndex].Data as ChoiceExtractDataResult;
                topResultData.OtherMatches = partialResults;
                results.Add(partialResults[topResultIndex]);
                partialResults.RemoveAt(topResultIndex);
            }
            else
            {
                results = partialResults;
            }

            return results;
        }

        private static int IndexOfToken(List<string> tokens, string token, int startPos)
        {
            if (tokens.Count <= startPos)
            {
                return -1;
            }

            return tokens.FindIndex(startPos, x => x == token);
        }

        private double MatchValue(IEnumerable<string> source, IEnumerable<string> match, int startPosition)
        {
            double matched = 0;
            var totalDeviation = 0;
            foreach (var token in match)
            {
                var pos = IndexOfToken(source.ToList(), token, startPosition);
                if (pos >= 0)
                {
                    var distance = matched > 0 ? pos - startPosition : 0;
                    if (distance <= config.MaxDistance)
                    {
                        matched++;
                        totalDeviation += distance;
                        startPosition = pos + 1;
                    }
                }
            }

            var score = 0.0;
            if (matched > 0 && (matched == match.Count() || config.AllowPartialMatch))
            {
                var completeness = matched / match.Count();
                var accuracy = completeness * (matched / (matched + totalDeviation));
                var initialScore = accuracy * (matched / source.Count());

                score = 0.4 + (0.6 * initialScore);
            }

            return score;
        }

        private IList<string> Tokenize(string text)
        {
            var tokens = new List<string>();
            var letters = UnicodeUtils.Letters(text);

            var token = string.Empty;
            foreach (var letter in letters)
            {
                if (UnicodeUtils.IsEmoji(letter))
                {
                    // Character is in a Supplementary Unicode Plane. This is where emoji live so
                    // we're going to just break each character in this range out as its own token.
                    tokens.Add(letter);
                    if (!string.IsNullOrWhiteSpace(token))
                    {
                        tokens.Add(token);
                        token = string.Empty;
                    }
                }
                else if (!(config.TokenRegex.IsMatch(letter) || string.IsNullOrWhiteSpace(letter)))
                {
                    token = token + letter;
                }
                else if (!string.IsNullOrWhiteSpace(token))
                {
                    tokens.Add(token);
                    token = string.Empty;
                }
            }

            if (!string.IsNullOrWhiteSpace(token))
            {
                tokens.Add(token);
                token = string.Empty;
            }

            return tokens;
        }
    }
}